text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+however, this needs to be merged and presented to the buildpack as one single
+directory. How that information is merged is outside of the scope of this
+document, but the directory structure defined in [Metadata
+Format](#metadata-format) guarantees that there will not be any duplicate
+dependency ids.
+
+This document does not specify how the dependency metadata folder should be
+provided to buildpacks but here are some possibilities:
+
+1. External. The metadata information can be managed outside of the buildpack
+   lifecycle. This allows for users to manually pull in the metadata they would
+   like when they would like it. The metadata can then be mapped into the build
+   container at the specified location using a volume mount to `pack build`.
+1. Via buildpack. We could add a Paketo buildpack to the beginning of buildpack
+   order groups that is responsible for pulling metadata and overrides
+   `BP_DEPENDENCY_METADATA` to point to its layer. The buildpack can check for
+   dependency metadata updates when it runs.
+1. Via builder. The builder could come with metadata included under
+   `/platform/deps/metadata`. This would likely require a builder with many
+   buildpacks to update very frequently though, so may not be the best idea for
+   the average case, however, it could be very useful as a way to distribute
+   dependencies in offline environments.
+1. Via the platform. Platforms may choose to offer enhanced functionality to
+   more easily distribute dependencies. In the end, the platform just needs to
+   ensure that the dependency metadata is available to the buildpack in the
+   required location.
+
+In addition to the flexibility of where the dependencies originate, this
+proposal also provides flexibility in how those dependencies are managed such
+as floating them so the latest versions are always available, pinning to a
+specific set of dependencies or even pinning and including the dependencies
+with the metadata.
+
+## Dependency Assets
+This proposal does not impact how dependencies are being distributed and it
+supports the current methods referring to remote locations, like upstream
+projects, or using the Deps Server. The dependency metadata just needs to point
+to the location from which the dependency should be fetched as is currently
+being done with the buildpack.toml dependency metadata.
+
+For builders, buildpacks, or platforms that would like to inject dependency
+assets directly into the build container, perhaps to support offline builds, we
+propose defining `BP_DEPENDENCY_BINARIES` which defaults to
+`/platform/deps/assets` as the location where the actual dependency assets
+should be located (again, the intent is to pilot and try out this or possibly
+other locations, eventually proposing an RFC with Cloud-Native buildpacks to
+standardize the location).
+
+In this way, one could have dependency metadata that uses `file://` URLs to
```

Comment:
I'm confused by this statement. I would have thought that the benefit of using this configurable path prefix is so that the dependency metadata doesn't have to include the full path.

I was thinking it would work like:

```
[[versions]]
...
uri = file://com/paketo/python/pip-1.2.3.tgz
```
```

which gets mapped to:

```
$BP_DEPENDENCY_BINARIES/com/paketo/python/pip-1.2.3.tgz
```

i.e. 

```
/platform/deps/assets/com/paketo/python/pip-1.2.3.tgz
```

As it currently reads, the proposal sounds like the metadata toml file has to include the full path, which makes me question where the environment variable (`BP_DEPENDENCY_BINARIES`) is being used.

I'm sure I'm missing something, so maybe an example here would help?

Reply:
At present, there no magic with the root.

If you are injecting metadata and assets the you need to coordinate. If you are putting assets at `/foo` then Metadata must be set with paths to that location.

It is very basic. Easy to implement.

@ForestEckhardt And I were talking about this. It's hard to tell how this will play out so we're thinking that we keep things simple now.

If we need to add in a "base" for metadata urls we can do that without breaking things later. This could also be something that tooling manages so it's less of an issue that it might appear.

Obviously if someone feels strongly about needing more in this part of the RFC we can do that. Just wanted to start minimally and work up.

Reply:
I'd say that we should rather exclude the second environment variable, i.e. `BP_DEPENDENCY_BINARIES`, in that case.

Reply:
I think I agree wit @loewenstein - if the metadata has to be full paths then I'm missing the value of `BP_DEPENDENCY_BINARIES`. How would that environment variable be consumed?

Reply:
I see what you're saying. That seems unnecessary in this context. I'll likely remove this, but let me think about this a little more just to make sure I'm not forgetting something.